### ISSUE TITLE
CPU-heavy operation slowing training

### ISSUE THREAD

**Issue by ashpoolresearch** [thuml/Time-Series-Library#324]
While training on a larger dataset, an operation (most likely train_loss = np.average(train_loss) ) is severely impacting training speeds, taking 20x longer time than an actual epoch. Is there a way to make this more efficient?

**Comment by ashpoolresearch** [comment]
singled it out to `test_loss = self.vali(test_data, test_loader, criterion)` that's causing problems. for some reason it doesnt run on the gpu while everything else does?

**Comment by harbingerarbitrager422** [comment]
> singled it out to `test_loss = self.vali(test_data, test_loader, criterion)` that's causing problems. for some reason it doesnt run on the gpu while everything else does?

test data is being loaded with batch_size of 1, so if your dataset is huge, the testing phase is essentially a long `for` loop, which is going to be slow. It does run on GPU but with a batch_size of 1 I suppose the GPU memory footprint is too small for you to notice. To speed up the training, simply comment out this line. It shouldn't pose any issue since you do still perform inference on the validation set

**Comment by lizhuoq** [comment]
> > singled it out to `test_loss = self.vali(test_data, test_loader, criterion)` that's causing problems. for some reason it doesnt run on the gpu while everything else does?
> 
> test data is being loaded with batch_size of 1, so if your dataset is huge, the testing phase is essentially a long `for` loop, which is going to be slow. It does run on GPU but with a batch_size of 1 I suppose the GPU memory footprint is too small for you to notice. To speed up the training, simply comment out this line. It shouldn't pose any issue since you do still perform inference on the validation set

I think this is correct for very large datasets, as it indeed wastes time and the `test_loss` during the training phase, besides being logged, serves no purpose, so I also agree with commenting out this line of code.

**Comment by wuhaixu2016** [comment]
Hi, the comment of @harbingerarbitrager422 is correct. We have updated this.
